docker/init_or_update_env.rb

```diff
@@ -0,0 +1,20 @@
+require 'set'
+
+# verify existing env vars
+existing_env_vars = Set.new(File.read("/.env").split)
+
+# hardcoded list of env vars we require
+required_env_vars = Set.new(["SECRET_KEY_BASE",
+"ENCRYPTION_PRIMARY_KEY",
+"ENCRYPTION_DETERMINISTIC_KEY",
+"ENCRYPTION_KEY_DERIVATION_SALT"])
+
+# set operation to get only env vars we need that don't exist yet
+new_env_var_keys = required_env_vars - existing_env_vars
+
+for new_env_var_key in new_env_var_keys do
+    # TODO: generate the default
```

Comment:
@mockdeep I'm struggling to conceptualize how to generate the env vars here. Am I on the wrong track? This feels a little primitive to do this set operation, and I seem to remember something about dotenv/erb that would help us here.

Reply:
Hmm, so we want to add any missing variables, and keep all of the existing ones, right? I assume we want to ignore anything that is already in the environment, whether in `.env` or elsewhere. So maybe we should load in the Rails environment, then grab just the missing keys and append those to the `.env` file. Something like:

```rb
require_relative "../config/environment"

required_env = {
  "SECRET_KEY_BASE" => `openssl rand -hex 64`.strip,
  "ENCRYPTION_PRIMARY_KEY" => `openssl rand -hex 64`.strip,
  "ENCRYPTION_DETERMINISTIC_KEY" => `openssl rand -hex 64`.strip,
  "ENCRYPTION_KEY_DERIVATION_SALT" => `openssl rand -hex 64`.strip,
}

required_env.each do |key, value|
  next if ENV[key].present?

  File.open(".env", "a") { |file| file << "#{key}=#{value}\n"
end
```

Reply:
I don't know why this wasn't obvious to me, it's like my brain turned to mush as soon as Ruby got involved. Thanks lol. I didn't know you could use `openssl rand` like that within Ruby, that's neat.

Reply:
Alright, now we're cooking! I was able to generate the `.env` file as discussed - we `touch .env`, then run `docker compose up stringer-setup`, and it writes any required-but-not-already-existing env vars to the env file in the host machine. Let me know if this Ruby script looks good? A couple of comments:
- I had to remove the `../config/environment` requirement because if the application was initialized without the required keys, it crashed
- I used `has_key?` instead of `present?` but I'm not sure what the difference is (if any). Is one more idiomatic than the other?
- I think the `DATABASE_URL` has to be added separately from the other vars, because it depends on the vars being generated. Unless we're able to reference hash keys/values that were already introduced, but before the hash was "finished"?
- For the `DATABASE_URL`, we need to get the actual values of the environment if they exist, so I used some ternaries to declare the db, username, password etc. Is this idiomatic? I saw something like `ENV[key] || default` as well that we could use, but that seems to have edge cases around nil/false. (maybe there's a different way of fetching a key from a hash with a default if it's not present, too?)